File: primes_sieve.cs
using System;
using System.Collections;
using System.Collections.Generic;
 
namespace SieveOfEratosthenes
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine ("C# - Sieve");
            int minprime = 0, maxprime = 20000000;
            var watch = System.Diagnostics.Stopwatch.StartNew();
            var primelist = GetAllPrimesLessThan(maxprime);
            //foreach (int prime in primelist)
            //{
            //    Console.WriteLine(prime);
            //}
            watch.Stop();
            Console.WriteLine ("Number of primes in the interval [{0},{1}]: {2}",minprime,maxprime,primelist.Count);
            Console.WriteLine ("Time elapsed: {0} seconds",watch.ElapsedMilliseconds/1000.0);
        }
 
        private static List<int> GetAllPrimesLessThan(int maxPrime)
        {
            var primes = new List<int>();
            var maxSquareRoot = (int)Math.Sqrt(maxPrime);
            var eliminated = new BitArray(maxPrime + 1);
 
            for (int i = 2; i <= maxPrime; ++i)
                if (!eliminated[i])
                {
                    primes.Add(i);
                    if (i <= maxSquareRoot)
                        for (int j = i * i; j <= maxPrime; j += i)
                            eliminated[j] = true;
                }
            return primes;
        }
    }
}

File: primes_sieve.cpp
#include <iostream>
#include <cstring>
#include <math.h>
#include <time.h>
#include <omp.h>

using namespace std;

long diff_nano(struct timespec *start, struct timespec *end)
{
    return ( (end->tv_sec * 1000000000 ) + (end->tv_nsec) -
             (start->tv_sec * 1000000000) - (start->tv_nsec) );
}

int runEratosthenesSieve(int upperBound) 
{
    int m, k;
    int upperBoundSquareRoot = (int)sqrt((double)upperBound);
    bool *isPrime = new bool[upperBound + 1];
    memset(isPrime, 1, sizeof(bool) * (upperBound + 1));
    int count = 0 ;

    int iCPU = omp_get_num_procs();
    omp_set_num_threads(iCPU);

    for (m = 2; m <= upperBoundSquareRoot; m++) 
        if (isPrime[m])
            #pragma omp parallel for private(k) shared(isPrime)
            for (k = m * m; k <= upperBound; k += m) 
                isPrime[k] = false;
                
 
    #pragma omp parallel for private(k) shared(isPrime) reduction(+:count) 
    for (m = 2; m <= upperBound; m++)
        if (isPrime[m])
            count++;
    
    delete [] isPrime;
    return count;
}

int main(int argc, char* argv[])
{
    struct timespec tstart, tend;

    int minprime = 2, maxprime = 20000000;

    std::cout << "C++ Sieve of Eratosthenes\n";

    clock_gettime(CLOCK_MONOTONIC, &tstart);
    int count = runEratosthenesSieve(maxprime);
    clock_gettime(CLOCK_MONOTONIC, &tend);

    double time_spent = diff_nano(&tstart, &tend) / 1000000.;

    std::cout << "Number of primes in the interval ["
              << minprime << "," << maxprime << "]: " << count
              << "\nTime elapsed: " << time_spent << " ms.\n";

    return 0;
}


